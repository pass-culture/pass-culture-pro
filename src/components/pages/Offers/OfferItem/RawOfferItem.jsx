import classnames from 'classnames'
import get from 'lodash.get'
import { Icon, pluralize } from 'pass-culture-shared'
import React, { Component } from 'react'
import Dotdotdot from 'react-dotdotdot'
import { NavLink } from 'react-router-dom'
import { requestData } from 'redux-saga-data'

import Price from 'components/layout/Price'
import Thumb from 'components/layout/Thumb'
import { offerNormalizer } from 'utils/normalizers'

class RawOfferItem extends Component {
  onDeactivateClick = () => {
    const { dispatch, offer } = this.props
    const { id, isActive } = offer || {}

    dispatch(
      requestData({
        apiPath: `/offers/${id}`,
        body: {
          isActive: !isActive,
        },
        isMergingDatum: true,
        isMutatingDatum: true,
        isMutaginArray: false,
        method: 'PATCH',
        normalizer: offerNormalizer,
      })
    )
  }

  buildNumberOfParticipantsLabel = (groupSizeMin, groupSizeMax) => {
    return groupSizeMin === groupSizeMax
      ? `${groupSizeMin}`
      : `${groupSizeMin} - ${groupSizeMax}`
  }

  buildNumberOfParticipantsTitle = (groupSizeMin, groupSizeMax) => {
    const groupLabel =
      groupSizeMin === groupSizeMax
        ? `minimum ${pluralize(groupSizeMin, 'personnes')}`
        : `entre ${groupSizeMin} et ${groupSizeMax} personnes`

    return groupSizeMin > 0 ? groupLabel : null
  }

  buildProductNavLinkLabel = (offer, stockSize) => {
    if (offer.isThing) {
      return `${stockSize} prix`
    }

    if (offer.isEvent) {
      return pluralize(stockSize, 'date')
    }
  }

  getThumbUrl(mediations) {
    const emptyUrl = ''

    return mediations
      ? mediations.length > 0
        ? mediations[0].thumbUrl
        : emptyUrl
      : emptyUrl
  }

  render() {
    const {
      aggregatedStock,
      location: { search },
      maxDate,
      mediations,
      offer,
      product,
      stocks,
      offerTypeLabel,
      offerer,
      stockAlertMessage,
      venue,
    } = this.props

    const { isNew } = offer || {}
    const { groupSizeMin, groupSizeMax, priceMin, priceMax } =
      aggregatedStock || {}
    const { name } = product || {}
    const thumbUrl = this.getThumbUrl(mediations)

    const numberOfMediations = get(mediations, 'length')
    const remainingStockQuantity = get(stocks, 'length')

    return (
      <li
        className={classnames('offer-item', {
          active: offer.isActive,
          product,
        })}>
        <Thumb alt="offre" src={thumbUrl} />
        <div className="list-content">
          <NavLink
            className="name"
            to={`/offres/${offer.id}${search}`}
            title={name}>
            <Dotdotdot clamp={1}>{name}</Dotdotdot>
          </NavLink>
          <ul className="infos">
            <li className="is-uppercase">{offerTypeLabel}</li>
            <li>
              <span className="label">Structure : </span>
              {offerer && offerer.name}
            </li>
            <li>
              <span className="label">Lieu : </span>
              {(venue && venue.publicName) || venue.name}
            </li>
          </ul>
          <ul className="infos">
            {isNew && (
              <li>
                <div className="recently-added" />
              </li>
            )}
            <li
              title={this.buildNumberOfParticipantsTitle(
                groupSizeMin,
                groupSizeMax
              )}>
              {groupSizeMin === 1 && <Icon svg="picto-user" />}
              {groupSizeMin > 1 && (
                <div>
                  <Icon svg="picto-group" />,{' '}
                  <p>
                    {this.buildNumberOfParticipantsLabel(
                      groupSizeMin,
                      groupSizeMax
                    )}
                  </p>
                </div>
              )}
            </li>
            <li>
              <NavLink
                className="has-text-primary"
                to={`/offres/${offer.id}?gestion`}>
                {this.buildProductNavLinkLabel(offer, remainingStockQuantity)}
              </NavLink>
            </li>
            <li>{maxDate && `jusqu'au ${maxDate.format('DD/MM/YYYY')}`}</li>
            {stockAlertMessage && <li>{stockAlertMessage}</li>}
            <li>
              {priceMin === priceMax ? (
                <Price value={priceMin || 0} />
              ) : (
                <span>
                  <Price value={priceMin} /> - <Price value={priceMax} />
                </span>
              )}
            </li>
          </ul>
          <ul className="actions">
            <li>
              <NavLink
                to={`/offres/${offer.id}${
                  numberOfMediations ? '' : `/accroches/nouveau${search}`
                }`}
                className={`button addMediations is-small ${
                  numberOfMediations ? 'is-secondary' : 'is-primary is-outlined'
                }`}>
                <span className="icon">
                  <Icon svg="ico-stars" />
                </span>
                <span>
                  {get(mediations, 'length')
                    ? 'Accroches'
                    : 'Ajouter une Accroche'}
                </span>
              </NavLink>
            </li>
            <li>
              <NavLink
                to={`/offres/${offer.id}`}
                className="button is-secondary is-small edit-link">
                <Icon svg="ico-pen-r" />
                Modifier
              </NavLink>

              <button
                className="button is-secondary is-small activ-switch"
                onClick={this.onDeactivateClick}>
                {offer.isActive ? (
                  <span>
                    <Icon svg="ico-close-r" />
                    DÃ©sactiver
                  </span>
                ) : (
                  'Activer'
                )}
              </button>
            </li>
          </ul>
        </div>
      </li>
    )
  }
}

RawOfferItem.defaultProps = {
  maxDescriptionLength: 300,
}
export default RawOfferItem
