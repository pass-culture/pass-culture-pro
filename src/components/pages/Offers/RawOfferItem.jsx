import { Icon, pluralize, requestData } from 'pass-culture-shared'
import classnames from 'classnames'
import get from 'lodash.get'
import moment from 'moment'
import React, { Component } from 'react'
import Dotdotdot from 'react-dotdotdot'
import { NavLink } from 'react-router-dom'

import Price from '../../layout/Price'
import Thumb from '../../layout/Thumb'

import { offerNormalizer } from '../../../utils/normalizers'

class RawOfferItem extends Component {
  onDeactivateClick = event => {
    const { dispatch, offer } = this.props
    const { id, isActive } = offer || {}
    dispatch(
      requestData('PATCH', `offers/${id}`, {
        body: {
          isActive: !isActive,
        },
        normalizer: offerNormalizer,
        isMergingDatum: true,
        isMutatingDatum: true,
        isMutaginArray: false,
      })
    )
  }

  render() {
    const {
      aggregatedStock,
      event,
      eventOccurrences,
      location: { search },
      maxDate,
      mediations,
      offer,
      stocks,
      thing,
      thumbUrl,
      offerTypeLabel,
      offerrer,
      venue,
    } = this.props

    const { isNew } = offer || {}
    const { available, groupSizeMin, groupSizeMax, priceMin, priceMax } =
      aggregatedStock || {}
    const { name, createdAt } = event || thing || {}

    const mediationsLength = get(mediations, 'length')

    return (
      <li
        className={classnames('offer-item', {
          active: offer.isActive,
          event,
          thing,
        })}>
        <Thumb alt="offre" src={thumbUrl} />
        <div className="list-content">
          <NavLink
            className="name"
            to={`/offres/${offer.id}${search}`}
            title={name}>
            <Dotdotdot clamp={1}>{name}</Dotdotdot>
          </NavLink>
          <ul className="infos">
            <li className="is-uppercase">{offerTypeLabel}</li>
            <li>
              <span className="label">Structure :</span>
              {offerrer && offerrer.name}
            </li>
            <li>
              <span className="label">Lieu :</span> {venue && venue.name}
            </li>
          </ul>
          <ul className="infos">
            {isNew && (
              <li>
                <div className="recently-added" />
              </li>
            )}
            {false && moment(createdAt).isAfter(moment().add(-1, 'days')) && (
              <li>
                <div className="recently-added" />
              </li>
            )}
            <li
              title={
                groupSizeMin > 0
                  ? groupSizeMin === groupSizeMax
                    ? `minimum ${pluralize(groupSizeMin, 'personnes')}`
                    : `entre ${groupSizeMin} et ${groupSizeMax} personnes`
                  : undefined
              }>
              {
                // DISABLE GROUP SIZE AMBIGUITY FOR THE MOMENT
                /*groupSizeMin === 0 && (
                <div>
                  <Icon svg="picto-user" /> {'ou '} <Icon svg="picto-group" />
                </div>
                )*/
              }
              {groupSizeMin === 1 && <Icon svg="picto-user" />}
              {groupSizeMin > 1 && (
                <div>
                  <Icon svg="picto-group" />,{' '}
                  <p>
                    {groupSizeMin === groupSizeMax
                      ? groupSizeMin
                      : `${groupSizeMin} - ${groupSizeMax}`}
                  </p>
                </div>
              )}
            </li>
            <li>
              <NavLink
                className="has-text-primary"
                to={`/offres/${offer.id}?gestion`}>
                {event
                  ? pluralize(get(eventOccurrences, 'length'), 'dates')
                  : get(stocks, 'length') + ' prix'}
              </NavLink>
            </li>
            <li>{maxDate && `jusqu'au ${maxDate.format('DD/MM/YYYY')}`}</li>
            <li>{available ? `encore ${available} places` : '0 place'} </li>
            <li>
              {priceMin === priceMax ? (
                <Price value={priceMin || 0} />
              ) : (
                <span>
                  <Price value={priceMin} /> - <Price value={priceMax} />
                </span>
              )}
            </li>
          </ul>
          <ul className="actions">
            <li>
              <NavLink
                to={`/offres/${offer.id}${
                  mediationsLength ? '' : `/accroches/nouveau${search}`
                }`}
                className={`button addMediations is-small ${
                  mediationsLength ? 'is-secondary' : 'is-primary is-outlined'
                }`}>
                <span className="icon">
                  <Icon svg="ico-stars" />
                </span>
                <span>
                  {get(mediations, 'length')
                    ? 'Accroches'
                    : 'Ajouter une Accroche'}
                </span>
              </NavLink>
            </li>
            <li>
              <NavLink
                to={`/offres/${offer.id}`}
                className="button is-secondary is-small edit-link">
                <Icon svg="ico-pen-r" />
                Modifier
              </NavLink>

              <button
                className="button is-secondary is-small activ-switch"
                onClick={this.onDeactivateClick}>
                {offer.isActive ? (
                  <span>
                    <Icon svg="ico-close-r" />
                    DÃ©sactiver
                  </span>
                ) : (
                  'Activer'
                )}
              </button>
            </li>
          </ul>
        </div>
      </li>
    )
  }
}

RawOfferItem.defaultProps = {
  maxDescriptionLength: 300,
}
export default RawOfferItem
