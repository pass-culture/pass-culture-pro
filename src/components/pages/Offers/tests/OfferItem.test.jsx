import React from 'react'
import { shallow } from 'enzyme'
import RawOfferItem from '../RawOfferItem'
import mockedOffers from './offersMock'

import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'

const middlewares = []
const mockStore = configureStore(middlewares)

describe('snapshot', () => {
  it('should match snapshot', () => {
    // given
    const initialState = {}
    const store = mockStore(initialState)
    const dispatchMock = jest.fn()
    const initialProps = {
      dispatch: dispatchMock,
      offer: mockedOffers[0],
      location: {
        search: '?orderBy=offer.id+desc',
      },
    }

    // when
    const wrapper = shallow(
      <Provider store={store}>
        <RawOfferItem {...initialProps} />
      </Provider>
    )

    // then
    expect(wrapper).toBeDefined()
    expect(wrapper).toMatchSnapshot()
    dispatchMock.mockClear()
  })
})
