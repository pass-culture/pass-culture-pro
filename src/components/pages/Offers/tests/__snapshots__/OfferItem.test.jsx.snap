// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
>
    <RawOfferItem
        dispatch={[Function]}
        location={
            Object {
                "search": "?orderBy=offer.id+desc",
              }
        }
        maxDescriptionLength={300}
        offer={
            Object {
                "bookingEmail": "booking.email@test.com",
                "dateCreated": "2019-02-25T09:50:10.735519Z",
                "dateModifiedAtLastProvider": "2019-02-25T09:50:31.881297Z",
                "eventId": null,
                "eventOccurrencesIds": Array [],
                "id": "M4",
                "idAtProviders": null,
                "isActive": true,
                "lastProviderId": null,
                "mediationsIds": Array [
                  "HA",
                ],
                "modelName": "Offer",
                "stocksIds": Array [
                  "MQ",
                ],
                "thingId": "EY",
                "venueId": "CU",
              }
        }
    />
</Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "dispatch": [Function],
      "location": Object {
        "search": "?orderBy=offer.id+desc",
      },
      "maxDescriptionLength": 300,
      "offer": Object {
        "bookingEmail": "booking.email@test.com",
        "dateCreated": "2019-02-25T09:50:10.735519Z",
        "dateModifiedAtLastProvider": "2019-02-25T09:50:31.881297Z",
        "eventId": null,
        "eventOccurrencesIds": Array [],
        "id": "M4",
        "idAtProviders": null,
        "isActive": true,
        "lastProviderId": null,
        "mediationsIds": Array [
          "HA",
        ],
        "modelName": "Offer",
        "stocksIds": Array [
          "MQ",
        ],
        "thingId": "EY",
        "venueId": "CU",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "dispatch": [Function],
        "location": Object {
          "search": "?orderBy=offer.id+desc",
        },
        "maxDescriptionLength": 300,
        "offer": Object {
          "bookingEmail": "booking.email@test.com",
          "dateCreated": "2019-02-25T09:50:10.735519Z",
          "dateModifiedAtLastProvider": "2019-02-25T09:50:31.881297Z",
          "eventId": null,
          "eventOccurrencesIds": Array [],
          "id": "M4",
          "idAtProviders": null,
          "isActive": true,
          "lastProviderId": null,
          "mediationsIds": Array [
            "HA",
          ],
          "modelName": "Offer",
          "stocksIds": Array [
            "MQ",
          ],
          "thingId": "EY",
          "venueId": "CU",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
